request.go

```diff
@@ -124,3 +128,49 @@ func detectEncoding(opts RequestOptions) encoding.Encoding {
 
 	return encoding.JSON
 }
+
+// prepares the request by injecting metadata, applying plugin-based transport middleware,
+// before finally adding any user-provided override values
+func prepareRequest(req *transport.Request, headers map[string]string, opts Options) (*transport.Request, error) {
+	reqWithMeta := mutateRequestWithMetadata(req, opts.TOpts)
```

Comment:
nit: We don't need to be this verbose with the operations. The operations
being executed are pretty trivial so it's preferable not to jump around, which
is especially difficult with the long similar-looking names.

Can you perform the operations inline? You can group blocks of code with
comments to serve the same purpose as these names.

```go
// Add request metadata
req.TargetService = ...

// Apply middleware
ctx, cancel := ...
defer cancel()
req, err := transportmiddleware.Apply(ctx, req)
if err != nil {
  return nil, err
}

// Apply command line arguments
req.Headers = ...
req.TransportHeaders = ...
```

I understand you have tests for the three different functions. I would
recommend collapsing those tests so that you're testing prepareRequest
directly. The reason for this being that we do want to test whether CLI
overrides actually override the default metadata and the middleware.

---

If you strongly prefer to keep the functions, please rename them to something
like, `addRequestMetadata`, `applyMiddleware`, and `applyCLIOverrides`
respectively.